### PR TITLE
Preparation for Updater code

### PR DESCRIPTION
**Pull request by apelisse**
- Change conflict creation for convenience,
- Initialize some variables and fixes to the test framework,
- Diff test from nothing to something
- Add IsSame on comparison for convenience,

### CODE REVIEW

File: merge/conflict.go [hunk 1]
```diff
@@ -74,8 +74,8 @@ func (conflicts Conflicts) Error() string {
 	return strings.Join(messages, "\n")
 }
 
-// NewFromSets creates a list of conflicts error from a map of owner to set of fields.
-func NewFromSets(sets map[string]*fieldpath.Set) Conflicts {
+// NewFromOwners creates a list of conflicts error from a map of owner to set of fields.
+func NewFromOwners(sets Owners) Conflicts {
```
[the remark this answers is not in the document]
Comment (apelisse) [reply merge/conflict.go]: For now, we sort of have owners everywhere. I'm keeping it consistent "for now", we can decide to go with manager and change it everywhere.

File: merge/conflict.go [hunk 2]
```diff
@@ -74,8 +74,8 @@ func (conflicts Conflicts) Error() string {
 	return strings.Join(messages, "\n")
 }
 
-// NewFromSets creates a list of conflicts error from a map of owner to set of fields.
-func NewFromSets(sets map[string]*fieldpath.Set) Conflicts {
+// NewFromOwners creates a list of conflicts error from a map of owner to set of fields.
```
[the remark this answers is not in the document]
Comment (apelisse) [reply merge/conflict.go]: Makes sense.

File: typed/typed.go [hunk 3]
```diff
@@ -143,6 +153,12 @@ func (tv TypedValue) Compare(rhs TypedValue) (c *Comparison, err error) {
 }
 
 func merge(lhs, rhs TypedValue, rule, postRule mergeRule) (TypedValue, error) {
+	if !rhs.initialized() {
```
[the remark this answers is not in the document]
Comment (apelisse) [reply typed/typed.go]: > This changes behavior from a typed-but-empty object--the whole tree is still walked for those.

Does it? initialized only matches entirely non-initialized object that don't have typeref or schema. typed-but-empty objects are considered initialized and this won't change the behavior

[the remark this answers is not in the document]
Comment (apelisse) [reply typed/typed.go]: I'm going to add a test for that, but so far none of the tests have been impacted by this change (we should probably add more to be extra sure).

**Comment by apelisse** [comment]
Updated.

File: typed/typed.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (lavalamp) [reply typed/typed.go]: I'm claiming if you send a real object with a typed-but-empty object, you'll find all the fields are walked and the compare function detects everything as added or deleted. But if you send with an uninitialized object, it would not list anything as added/deleted.

For that reason I don't think this should accept uninitialized objects, I think we should make it easy to make objects initialized with the right type.

File: typed/merge_test.go [hunk 4]
```diff
@@ -378,3 +386,57 @@ func TestMerge(t *testing.T) {
...
+		t.Fatalf("got validation errors: %v", err)
+	}
+	if !reflect.DeepEqual(got.AsValue().ToUnstructured(true), init.AsValue().ToUnstructured(true)) {
+		t.Errorf("Expected\n%v\nbut got\n%v\n",
+			init.AsValue().HumanReadable(), got.AsValue().HumanReadable(),
+		)
+	}
+
+	// Change the order, uninit is init, init is uninit, same result.
+	got, err = init.Merge(uninit)
```
Comment (lavalamp) [inline typed/merge_test.go]: If you test Compare instead of Merge it'll show the thing I'm complaining about :)

Comment (lavalamp) [reply typed/merge_test.go]: Like the Comparison.IsSame would return true...

**Comment by apelisse** [comment]
OK, I've dropped the uninitialized merge commit altogether.

**Comment by kwiesmueller** [comment]
Looks good to me, but why is handling a nil owners object part of the test framework? Wont' this get handled inside the updater?

**Comment by apelisse** [comment]
> why is handling a nil owners object part of the test framework? Wont' this get handled inside the updater?

Yeah, you're right, I could move that in the updater

**Comment by lavalamp** [comment]
/lgtm
/approve

Thanks!
